File: java/RobotCommon/src/robot/common/Direction.java
/*
 * Direction
 *
 * Copyright (C) 2010  Jaroslav Merxbauer
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package robot.common;

import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.Map.Entry;

/**
 *
 * @author Jaroslav Merxbauer
 * @version %I% %G%
 */
public enum Direction {

    North, West, South, East, Unknown;

    protected static final List<Direction> directionRotationOrder;
    protected static final EnumMap<Direction, Vector> dirToVec;
    static {
        directionRotationOrder = Arrays.asList(new Direction[] {Direction.North, Direction.West, Direction.South, Direction.East});
        dirToVec = new EnumMap<Direction, Vector>(Direction.class);
        dirToVec.put(Direction.North,     new Vector( 0, 1));
        dirToVec.put(Direction.East,      new Vector( 1, 0));
        dirToVec.put(Direction.South,     new Vector(-1, 0));
        dirToVec.put(Direction.West,      new Vector( 1, 0));
        dirToVec.put(Direction.Unknown,   new Vector( 0, 0));
    }

    public static Direction getNextDirection(Direction current) {
        int directionIndex = directionRotationOrder.indexOf(current);
        return directionRotationOrder.get((directionIndex + 1) % 4);
    }

    public static Vector toVector(Direction direction) {
        return dirToVec.get(direction);
    }

    public static Direction fromVector(Vector vector) {
        for (Entry<Direction, Vector> e : dirToVec.entrySet()) {
            if (e.getValue().equals(vector)) {
                return e.getKey();
            }
        }
        return null;
    }

}

File: cpp/Robot/RobotCommon/Direction.cpp
#include "Direction.h"
#include "Vector.h"
#include <algorithm>

using RobotCommon::Direction;
using RobotCommon::DirectionUtils;
using RobotCommon::Vector;
using std::map;
using std::copy;
using std::back_inserter;
using std::begin;
using std::end;
using std::next;
using std::find;

bool DirectionUtils::initialized = false;
list<Direction> DirectionUtils::directionRotationOrder;
map<Direction, Vector> DirectionUtils::dirToVec;
map<Vector, Direction> DirectionUtils::vecToDir;

void DirectionUtils::ensureInitialized(void)
{
    if (!initialized)
    {
        initialize();
    }
}

void DirectionUtils::initialize(void)
{
    Direction dirs[] = {North, West, South, East};
    copy(begin(dirs), end(dirs), back_inserter(directionRotationOrder));

    dirToVec[North]   = Vector( 0, 1);
    dirToVec[East]    = Vector( 1, 0);
    dirToVec[South]   = Vector( 0,-1);
    dirToVec[West]    = Vector(-1, 0);
    dirToVec[Unknown] = Vector( 0, 0);

    vecToDir[Vector( 0, 1)] = North;
    vecToDir[Vector( 1, 0)] = East;
    vecToDir[Vector( 0,-1)] = South;
    vecToDir[Vector(-1, 0)] = West;
    vecToDir[Vector( 0, 0)] = Unknown;

    initialized = true;
}

Direction DirectionUtils::GetNextDirection(Direction current)
{
    ensureInitialized();

    auto itNext = next(find(begin(directionRotationOrder), end(directionRotationOrder), current));
    return (itNext != end(directionRotationOrder) ? *itNext : *(begin(directionRotationOrder)));
}

Vector DirectionUtils::ToVector(Direction direction)
{
    ensureInitialized();

    return dirToVec[direction];
}

Direction DirectionUtils::FromVector(Vector vector)
{
    ensureInitialized();

    return vecToDir[vector];
}

wstring DirectionUtils::ToString(Direction direction)
{
    switch (direction)
    {
    case North:
        return L"North";
    case South:
        return L"South";
    case East:
        return L"East";
    case West:
        return L"West";
    default:
        return L"Unknown";
    };
}